### DISCUSSION TITLE
Problem with serialization custom property

### DISCUSSION THREAD

**Discussion by Pro100proff** [relayrides/pushy#973] [Q&A]
I have encountered with problem when I wanted to add custom property as my own object. I expected json object like this:

```json
{
       "aps": {
        "alert": {
            "body": "body text",
            "title": "title text"
        },
        "sound": "default",
        "badge": 1
    },
    "data": {
        "deeplink": "rupodari://wallet"
    }
}
```

In reality, I got:

```json
{
       "aps": {
        "alert": {
            "body": "body text",
            "title": "title text"
        },
        "sound": "default",
        "badge": 1
    },
    "data": "Data(deeplink=\"rupodari:\/\/wallet\")"
}
```

I checked the implementation and realised that it used toString method for the serialization custom properties. 
Is there work around to solve it?

**Comment by jchambers** [comment]
Well, the docs say:

> In addition to the Apple-defined keys, you may add custom keys to your payload to deliver small amounts of data to your app, notification service app extension, or notification content app extension. Your custom keys must have values with primitive types, such as dictionary, array, string, number, or Boolean. Custom keys are available in the [userInfo](https://developer.apple.com/documentation/usernotifications/unnotificationcontent/1649869-userinfo) dictionary of the UNNotificationContent object delivered to your app.

So I guess this is just a bug. I'll open an issue.

**Reply by jchambers** [reply]
I think I've misunderstood the situation and have un-marked this as the answer.

**Comment by jchambers** [comment]
Cross-posting from #974:

> @Pro100proff I've looked into this a little more and I think I misunderstood your original report. To be more precisely state what I think is happening here, I understand that:
> 
> - You are using a `SimpleApnsPayloadBuilder`
> - You are trying to set a custom property, and that value of that property is not a Java primitive, array of primitives, a `Collection`, or a `Map`
> 
> If I'm understanding you correctly, then this behavior is expected and is, in fact, documented: https://github.com/jchambers/pushy/blob/77453f3aa07482e7cf22ab626a324ff680d789ad/pushy/src/main/java/com/eatthepath/pushy/apns/util/ApnsPayloadBuilder.java#L690-L716
> 
> If you want to change how custom properties are serialized, you'll need to use a different `ApnsPayloadBuilder` implementation. Pushy offers two as separate modules: [`GsonApnsPayloadBuilder`](https://github.com/jchambers/pushy/tree/master/gson-payload-builder) and [`JacksonApnsPayloadBuilder`](https://github.com/jchambers/pushy/tree/master/jackson-payload-builder).
> 
> That said, I think there's room to make the documentation clearer in this case and will open a pull request to do exactly that.

**Reply by Pro100proff** [reply]
Thanks a lot for your reply!
Could you explain, why you decided to put these serializers in separate libraries?

**Reply by jchambers** [reply]
> Could you explain, why you decided to put these serializers in separate libraries?

Some people want to use Gson. Some people want to use Jackson. Some people want to use neither. Moving payload builders that depend on third-party JSON libraries allows Pushy users to make a deliberate choice about which things they're adding to their classpath.
